lib/tree.rb

```diff
@@ -16,40 +16,161 @@ def initialize
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+    return current_node
+  end
+
+  # Time Complexity: ranges from O(log n) for a well-balanced tree to O(n) for a totally unbalanced tree
+  # Space Complexity: same as time complexity since we have to keep track of previous operations until we hit the base case
   def find(key)
-    raise NotImplementedError
+    find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # find helper
+  def find_helper(current_node, key)
+    
+    # base cases
+    return nil if current_node.nil?
+    return current_node.value if current_node.key == key
+
+    # if key sought is lower than current_node.key, move to the left
+    if key < current_node.key
+      find_helper(current_node.left, key)
+    else 
+      find_helper(current_node.right, key) 
+    end
+
+  end
+
+  # Time Complexity: O(n) because we need to traverse every node
+  # Space Complexity: also O(n) because we need to keep track of each stack call
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current_node, list)
+    # traversals for inorder: left-root-right (root)
+
+    # base case
+    return list if current_node.nil?
+
+    # recursive cases
+    inorder_helper(current_node.left, list)
+    list << { key: current_node.key, value: current_node.value }
+    inorder_helper(current_node.right, list)
+
+    return list
+
+  end 
+
+  # Time Complexity: O(n) because we need to traverse every node
+  # Space Complexity: also O(n) because we need to keep track of each stack call
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(current_node, list)
+    # traversals for preorder: root-left-right
+
+    # base case
+    return list if current_node.nil?
+
+    # recursive cases
+    list << { key: current_node.key, value: current_node.value }
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+
+  end
+
+  # Time Complexity: O(n) because we need to traverse every node
+  # Space Complexity: also O(n) because we need to keep track of each stack call
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(current_node, list)
+    # traversals for postorder: left-right-root
+
+    # base case
+    return list if current_node.nil?
+
+    # recursive cases
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list << { key: current_node.key, value: current_node.value }
+  end
+
+  # Time Complexity: O(n) since we are checking each node
+  # Space Complexity: Also O(n)
   def height
```

Comment:
The space complexity is O(n) if the tree is unbalanced and O(log n) if the tree is balanced.